File: p3.cpp
#include <iostream>
#include <string>
#include <iomanip>
#include <cmath>
#include "p3.h"
//g++ main.cpp p3.cpp -Wall -Wuninitialized -o a.out
//fails when copy or destruct is called
using namespace std;

Person::Person() {
    first = "";
    last = "";
    height = 0;
    weight = 0;
    prevHeight = nullptr;
    nextHeight = nullptr;
    nextWeight = nullptr;
}
Person::Person(std::string first, std::string last, int height, int weight) {
    this->first = first;
    this->last = last;
    this->height = height;
    this->weight = weight;
    score = 0;
    weightScore = 0;
    prevWeight = nullptr;
    prevHeight = nullptr;
    nextHeight = nullptr;
    nextWeight = nullptr;
}


// PersonList Class Definitions
PersonList::PersonList() {
    headHeightList = nullptr;
    headWeightList = nullptr;
    curr = nullptr;
    tailHeightList = nullptr;
    tailWeightList = nullptr;
    size = 0;
}

int PersonList::getSize()
{
    return this->size;
}

void PersonList::printByHeight(std::ostream& os) {
    //os << "Printing list by height:" << std::endl;
    
    // sort the list in descending order of height using insertion sort
    Person *sorted = nullptr;
    Person *curr = headHeightList;
    while (curr != nullptr) {
        Person *next = curr->nextHeight;
        if (sorted == nullptr || curr->height > sorted->height) {
            curr->nextHeight = sorted;
            sorted = curr;
        } else {
            Person *temp = sorted;
            while (temp->nextHeight != nullptr && curr->height < temp->nextHeight->height) {
                temp = temp->nextHeight;
            }
            curr->nextHeight = temp->nextHeight;
            temp->nextHeight = curr;
        }
        curr = next;
    }

    // print the sorted list
    curr = sorted;
    while (curr != nullptr) {
        os << curr->first << " " << curr->last << ": height=" << curr->height << ", weight=" << curr->weight << std::endl;
        curr = curr->nextHeight;
    }
}

void PersonList::printByWeight(ostream &os) {
    //os << "List ordered by weight:" << endl;
    Person* curr = this->tailWeightList;
    while (curr != nullptr) {
        os << curr->first << " " << curr->last << ": height=" << curr->height << ", weight=" << curr->weight << endl;
        curr = curr->prevWeight;
    }
}

//   DEFAULT
bool PersonList::add(std::string first, std::string last, int height, int weight) {
    // check if the person already exists in the list
    if (exists(first, last)) {
        return false;
    }

    // create a new person with the given information
    Person* newPerson = new Person(first, last, height, weight);

    // calculate the score for the new person based on their height and weight
    double score = height - sqrt(weight) / 10.0;

    // case 1: the list is empty
    if (headHeightList == nullptr) {
        headHeightList = newPerson;
        tailHeightList = newPerson;
        headWeightList = newPerson;
        tailWeightList = newPerson;
        size++;
        return true;
    }

    // find the position to insert the new person based on their score
    Person* curr = headHeightList;
    while (curr != nullptr) {
        double currScore = curr->height - sqrt(curr->weight) / 10.0;
        if (score > currScore) {
            // insert new person before the current person in the height list
            newPerson->nextHeight = curr;
            newPerson->prevHeight = curr->prevHeight;
            if (curr->prevHeight != nullptr) {
                curr->prevHeight->nextHeight = newPerson;
            } else {
                headHeightList = newPerson;
            }
            curr->prevHeight = newPerson;

            // insert new person before the current person in the weight list
            newPerson->nextWeight = curr;
            newPerson->prevWeight = curr->prevWeight;
            if (curr->prevWeight != nullptr) {
                curr->prevWeight->nextWeight = newPerson;
            } else {
                headWeightList = newPerson;
            }
            curr->prevWeight = newPerson;
            size++;
            return true;
        }
        curr = curr->nextHeight;
    }

    // add the new person to the end of the height list if their score is lower than all existing persons
    newPerson->prevHeight = tailHeightList;
    tailHeightList->nextHeight = newPerson;
    tailHeightList = newPerson;

    // add the new person to the end of the weight list if their score is higher than all existing persons
    newPerson->prevWeight = tailWeightList;
    tailWeightList->nextWeight = newPerson;
    tailWeightList = newPerson;

    size++;
    return true;
}


bool PersonList::exists(std::string first, std::string last) {
    Person* curr = headHeightList;
    while (curr != nullptr) {
        if (curr->first == first && curr->last == last) {
            return true;
        }
        curr = curr->nextHeight;
    }
    return false;
}


int PersonList::getHeight(string first, string last) 
{
    Person *curr = this->headHeightList;
    while (curr != nullptr) 
    {
        if (curr->first == first && curr->last == last) 
        {
            return curr->height;
        }
        curr = curr->nextHeight;
    }
    // if person not found
    return -1;
}


int PersonList::getWeight(string first, string last) 
{
    Person* curr = this->headWeightList;
    while (curr != nullptr) 
    {
        if (curr->first == first && curr->last == last) 
        {
            return curr->weight;
        }
        curr = curr->nextHeight;
    }
    return -1;
}



bool PersonList::remove(string first, string last) {
    Person *curr = this->headHeightList;
    Person *prev = nullptr;
    bool found = false;
    
    // Search for the node to remove in the height list
    while (curr != nullptr && !found) {
        if (curr->first == first && curr->last == last) {
            found = true;
            if (prev == nullptr) {
                // Removing the head node
                this->headHeightList = curr->nextHeight;
            } else {
                prev->nextHeight = curr->nextHeight;
            }
            if (curr->nextHeight != nullptr) {
                curr->nextHeight->prevHeight = prev;
            }
            if (tailHeightList == curr) {
                tailHeightList = prev;
            }
            curr->prevHeight = nullptr;
            curr->nextHeight = nullptr;
        } else {
            prev = curr;
        }
        curr = curr->nextHeight;
    }
    
    if (!found) {
        // Node not found in list
        return false;
    }
    
    // Search for the node to remove in the weight list
    curr = this->headWeightList;
    prev = nullptr;
    while (curr != nullptr) {
        if (curr->first == first && curr->last == last) {
            if (prev == nullptr) {
                // Removing the head node
                this->headWeightList = curr->nextWeight;
            } else {
                prev->nextWeight = curr->nextWeight;
            }
            if (curr->nextWeight != nullptr) {
                curr->nextWeight->prevWeight = prev;
            }
            if (tailWeightList == curr) {
                tailWeightList = prev;
            }
            delete curr;
            this->size--;
            return true;
        } else {
            prev = curr;
        }
        curr = curr->nextWeight;
    }
    
    // Node not found in weight list (should never happen)
    return false;
}


bool PersonList::updateName(std::string oldFirst, std::string oldLast, std::string newFirst, std::string newLast) {
    if (!exists(oldFirst, oldLast)) {
        return false;
    }
    Person* curr = headHeightList;
    while (curr != nullptr) {
        if (curr->first == oldFirst && curr->last == oldLast) {
            curr->first = newFirst;
            curr->last = newLast;
            return true;
        }
        curr = curr->nextHeight;
    }
    return false;
}

// DEFAULT for height
bool PersonList::updateHeight(std::string first, std::string last, int newHeight) {
    if (!exists(first, last)) {
        return false;
    }
    Person* curr = headHeightList;
    while (curr != nullptr) {
        if (curr->first == first && curr->last == last) {
            // remove the person from the list
            if (curr->prevHeight == nullptr) {
                headHeightList = curr->nextHeight;
            } else {
                curr->prevHeight->nextHeight = curr->nextHeight;
            }
            if (curr->nextHeight == nullptr) {
                tailHeightList = curr->prevHeight;
            } else {
                curr->nextHeight->prevHeight = curr->prevHeight;
            }
            // update the height and re-insert the person into the list
            curr->height = newHeight;
            double score = curr->height - sqrt(curr->weight) / 10.0;
            if (headHeightList == nullptr || score > headHeightList->height - sqrt(headHeightList->weight) / 10.0) {
                curr->prevHeight = nullptr;
                curr->nextHeight = headHeightList;
                if (headHeightList != nullptr) {
                    headHeightList->prevHeight = curr;
                } else {
                    tailHeightList = curr;
                }
                headHeightList = curr;
            } else {
                Person* temp = headHeightList;
                while (temp->nextHeight != nullptr && score <= temp->nextHeight->height - sqrt(temp->nextHeight->weight) / 10.0) {
                    temp = temp->nextHeight;
                }
                curr->prevHeight = temp;
                curr->nextHeight = temp->nextHeight;
                if (temp->nextHeight != nullptr) {
                    temp->nextHeight->prevHeight = curr;
                } else {
                    tailHeightList = curr;
                }
                temp->nextHeight = curr;
            }
            return true;
        }
        curr = curr->nextHeight;
    }
    return false;
}


bool PersonList::updateWeight(std::string first, std::string last, int newWeight) {
    Person *curr = headWeightList;
    while (curr != nullptr) {
        if (curr->first == first && curr->last == last) {
            // remove the person from the weight list
            if (curr->prevWeight == nullptr) {
                headWeightList = curr->nextWeight;
            } else {
                curr->prevWeight->nextWeight = curr->nextWeight;
            }
            if (curr->nextWeight == nullptr) {
                tailWeightList = curr->prevWeight;
            } else {
                curr->nextWeight->prevWeight = curr->prevWeight;
            }

            // update the weight and re-insert the person into the weight list
            curr->weight = newWeight;
            Person* temp = headWeightList;
            while (temp != nullptr && newWeight <= temp->weight) {
                temp = temp->nextWeight;
            }

            if (temp == headWeightList) {
                curr->prevWeight = nullptr;
                curr->nextWeight = headWeightList;
                headWeightList->prevWeight = curr;
                headWeightList = curr;
            } else if (temp == nullptr) {
                curr->prevWeight = tailWeightList;
                curr->nextWeight = nullptr;
                tailWeightList->nextWeight = curr;
                tailWeightList = curr;
            } else {
                curr->prevWeight = temp->prevWeight;
                curr->nextWeight = temp;
                temp->prevWeight->nextWeight = curr;
                temp->prevWeight = curr;
            }

            return true;
        }
        curr = curr->nextWeight;
    }
    return false;
}


//destruct
PersonList::~PersonList() {
    // Delete all nodes in headHeightList
    while (headHeightList != nullptr) {
        //cout << "Debug 1" << endl;
        Person* temp = headHeightList;
        //cout << "Debug 2" << endl;
        headHeightList = headHeightList->nextHeight;
        //cout << "Debug 3" << endl;
        delete temp;
        //cout << "Debug 4" << endl;
    }
   // Reset head and tail pointers to null
    //cout << "Debug 8" << endl;
    headHeightList = nullptr;
    headWeightList = nullptr;
    tailHeightList = nullptr;
    tailWeightList = nullptr;
}


const PersonList& PersonList::operator=(const PersonList& src) {
    // Check for self-assignment
    if (this == &src) {
        return *this;
    }

    // Call deepCopy to clear the current list and copy the contents of the source list
    deepCopy(src);

    return *this;
}


//copycon
PersonList::PersonList(const PersonList& src) {
    // Initialize variables
    headHeightList = nullptr;
    headWeightList = nullptr;
    tailHeightList = nullptr;
    tailWeightList = nullptr;
    size = 0;

    // Call deepCopy to get a deepcopy of the source list
    deepCopy(src);
}


PersonList& PersonList::deepCopy(const PersonList& src) {
    // Clear the current list
    Person* curr = headHeightList;
    while (curr != nullptr) {
        Person* temp = curr;
        curr = curr->nextHeight;
        delete temp;
    }
    headHeightList = nullptr;
    headWeightList = nullptr;
    tailHeightList = nullptr;
    tailWeightList = nullptr;
    size = 0;

    // Copy contents of the source list to the current list
    for (Person* curr = src.headHeightList; curr != nullptr; curr = curr->nextHeight) {
        // Use the add function to add the person, maintaining the correct order
        add(curr->first, curr->last, curr->height, curr->weight);
    }

    return *this;
}

File: p3.py
# p3.py
# Python translation of the C++ Person / PersonList project.
# Maintains two doubly linked lists: one sorted by height (desc), one by weight (desc).

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional, Iterable, Tuple, List, Any


@dataclass
class Person:
    first: str
    last: str
    height: int
    weight: int
    score: float = 0.0
    weightScore: float = 0.0

    # Pointers for height-sorted list
    prevHeight: Optional['Person'] = field(default=None, repr=False)
    nextHeight: Optional['Person'] = field(default=None, repr=False)

    # Pointers for weight-sorted list
    prevWeight: Optional['Person'] = field(default=None, repr=False)
    nextWeight: Optional['Person'] = field(default=None, repr=False)

    def key_height(self) -> Tuple[int, str, str]:
        # Sort primarily by height (desc), then by last/first for stability
        return (-self.height, self.last, self.first)

    def key_weight(self) -> Tuple[int, str, str]:
        # Sort primarily by weight (desc)
        return (-self.weight, self.last, self.first)

    def name_tuple(self) -> Tuple[str, str]:
        return (self.first, self.last)


class PersonList:
    def __init__(self) -> None:
        self.headHeightList: Optional[Person] = None
        self.tailHeightList: Optional[Person] = None
        self.headWeightList: Optional[Person] = None
        self.tailWeightList: Optional[Person] = None
        self.size: int = 0

    # ---- utilities -------------------------------------------------------
    def _find_by_name(self, first: str, last: str) -> Optional[Person]:
        cur = self.headHeightList
        while cur:
            if cur.first == first and cur.last == last:
                return cur
            cur = cur.nextHeight
        return None

    def exists(self, first: str, last: str) -> bool:
        return self._find_by_name(first, last) is not None

    def getSize(self) -> int:
        return self.size

    # ---- insert helpers (height / weight) --------------------------------
    def _insert_height(self, p: Person) -> None:
        if not self.headHeightList:
            self.headHeightList = self.tailHeightList = p
            p.prevHeight = p.nextHeight = None
            return
        cur = self.headHeightList
        while cur and p.key_height() > cur.key_height():
            cur = cur.nextHeight
        if cur is self.headHeightList:
            # insert at head
            p.prevHeight = None
            p.nextHeight = self.headHeightList
            self.headHeightList.prevHeight = p
            self.headHeightList = p
        elif cur is None:
            # insert at tail
            p.prevHeight = self.tailHeightList
            p.nextHeight = None
            self.tailHeightList.nextHeight = p
            self.tailHeightList = p
        else:
            # insert before cur (middle)
            p.prevHeight = cur.prevHeight
            p.nextHeight = cur
            assert cur.prevHeight is not None
            cur.prevHeight.nextHeight = p
            cur.prevHeight = p

    def _insert_weight(self, p: Person) -> None:
        if not self.headWeightList:
            self.headWeightList = self.tailWeightList = p
            p.prevWeight = p.nextWeight = None
            return
        cur = self.headWeightList
        while cur and p.key_weight() > cur.key_weight():
            cur = cur.nextWeight
        if cur is self.headWeightList:
            # insert at head
            p.prevWeight = None
            p.nextWeight = self.headWeightList
            self.headWeightList.prevWeight = p
            self.headWeightList = p
        elif cur is None:
            # insert at tail
            p.prevWeight = self.tailWeightList
            p.nextWeight = None
            self.tailWeightList.nextWeight = p
            self.tailWeightList = p
        else:
            # insert before cur (middle)
            p.prevWeight = cur.prevWeight
            p.nextWeight = cur
            assert cur.prevWeight is not None
            cur.prevWeight.nextWeight = p
            cur.prevWeight = p

    # ---- remove helpers (height / weight) --------------------------------
    def _unlink_height(self, p: Person) -> None:
        if p.prevHeight:
            p.prevHeight.nextHeight = p.nextHeight
        else:
            self.headHeightList = p.nextHeight
        if p.nextHeight:
            p.nextHeight.prevHeight = p.prevHeight
        else:
            self.tailHeightList = p.prevHeight
        p.prevHeight = p.nextHeight = None

    def _unlink_weight(self, p: Person) -> None:
        if p.prevWeight:
            p.prevWeight.nextWeight = p.nextWeight
        else:
            self.headWeightList = p.nextWeight
        if p.nextWeight:
            p.nextWeight.prevWeight = p.prevWeight
        else:
            self.tailWeightList = p.prevWeight
        p.prevWeight = p.nextWeight = None

    # ---- public API ------------------------------------------------------
    def add(self, first: str, last: str, height: int, weight: int) -> bool:
        if self.exists(first, last):
            return False
        p = Person(first, last, int(height), int(weight))
        self._insert_height(p)
        self._insert_weight(p)
        self.size += 1
        return True

    def remove(self, first: str, last: str) -> bool:
        p = self._find_by_name(first, last)
        if not p:
            return False
        self._unlink_height(p)
        self._unlink_weight(p)
        self.size -= 1
        return True

    def getHeight(self, first: str, last: str) -> int:
        p = self._find_by_name(first, last)
        return p.height if p else -1

    def getWeight(self, first: str, last: str) -> int:
        # search weight list (fast if many with same height)
        cur = self.headWeightList
        while cur:
            if cur.first == first and cur.last == last:
                return cur.weight
            cur = cur.nextWeight
        return -1

    def updateName(self, oldFirst: str, oldLast: str, newFirst: str, newLast: str) -> bool:
        p = self._find_by_name(oldFirst, oldLast)
        if not p:
            return False
        if (oldFirst, oldLast) != (newFirst, newLast) and self.exists(newFirst, newLast):
            # avoid duplicates
            return False
        p.first, p.last = newFirst, newLast
        return True

    def _reposition_after_height_change(self, p: Person) -> None:
        # unlink then reinsert in height list
        self._unlink_height(p)
        self._insert_height(p)

    def _reposition_after_weight_change(self, p: Person) -> None:
        self._unlink_weight(p)
        self._insert_weight(p)

    def updateHeight(self, first: str, last: str, newHeight: int) -> bool:
        p = self._find_by_name(first, last)
        if not p:
            return False
        p.height = int(newHeight)
        self._reposition_after_height_change(p)
        return True

    def updateWeight(self, first: str, last: str, newWeight: int) -> bool:
        p = self._find_by_name(first, last)
        if not p:
            return False
        p.weight = int(newWeight)
        self._reposition_after_weight_change(p)
        return True

    # ---- printing (does not mutate structure) ----------------------------
    def iter_by_height(self) -> Iterable[Person]:
        cur = self.headHeightList
        while cur:
            yield cur
            cur = cur.nextHeight

    def iter_by_weight(self) -> Iterable[Person]:
        cur = self.headWeightList
        while cur:
            yield cur
            cur = cur.nextWeight

    def printByHeight(self) -> str:
        lines = []
        for p in self.iter_by_height():
            lines.append(f"{p.first} {p.last}: height={p.height}, weight={p.weight}")
        return "\n".join(lines)

    def printByWeight(self) -> str:
        lines = []
        for p in self.iter_by_weight():
            lines.append(f"{p.first} {p.last}: height={p.height}, weight={p.weight}")
        return "\n".join(lines)

    # ---- copying ---------------------------------------------------------
    def deepCopy(self, src: 'PersonList') -> 'PersonList':
        # Clear current
        self.__init__()
        # Copy each person in height order to preserve both orders via add()
        for p in src.iter_by_height():
            self.add(p.first, p.last, p.height, p.weight)
        return self

    def copy(self) -> 'PersonList':
        new = PersonList()
        new.deepCopy(self)
        return new

    # Python dunder helpers
    def __len__(self) -> int:
        return self.size

    def __iter__(self) -> Iterable[Person]:
        return self.iter_by_height()

    def __repr__(self) -> str:
        return f"PersonList(size={self.size})"


if __name__ == "__main__":
    # Simple manual test / demo
    pl = PersonList()
    pl.add("Alice", "Zephyr", 65, 140)
    pl.add("Bob", "Yellow", 70, 180)
    pl.add("Carol", "Xavier", 62, 150)
    pl.add("Dave", "Williams", 70, 175)

    print("By height (desc):")
    print(pl.printByHeight(), "\n")

    print("By weight (desc):")
    print(pl.printByWeight(), "\n")

    print("Update Dave weight to 200...")
    pl.updateWeight("Dave", "Williams", 200)
    print(pl.printByWeight(), "\n")

    print("Remove Alice Zephyr...")
    pl.remove("Alice", "Zephyr")
    print(pl.printByHeight())
